### PR TITLE
[WIP] Update CTR prediction example with RandomForest workflow

### PR DESCRIPTION
**Pull request by takuti**
Treasure Data now supports the latest version of Hivemall. The update makes RandomForest more useful.

### CODE REVIEW

**Comment by myui** [comment]
@takuti `rescaling` and `amplify` is not required for RandomForests. Better to avoid them.

Just providing `-stratified_sampling` for unbalanced data.

**Comment by myui** [comment]
Also, RandomForest supports weighted sampling.

[Calibration](https://github.com/treasure-data/workflow-examples/pull/102/files#diff-d1f9494ec9a4bf896f4f796adaf9c965R22) does not work for RF prediction result since probability estimation of RandomForest is just by voting. 

Probability estimation of RandomForests is effective for AUC optimization but not good for logloss optimization.

**Comment by myui** [comment]
hivemall supports [classWeights](https://github.com/haifengl/smile/issues/163) for unbalanced data but I think this option is bad idea and unnecessary for randomforests.

https://github.com/apache/incubator-hivemall/blob/master/core/src/main/java/hivemall/smile/classification/RandomForestClassifierUDTF.java#L85

**Comment by takuti** [comment]
Yes. Don't worry, it's still WIP; I'm planning to eventually update queries as you suggested. Now I'm aggressively modifying my local file for the sake of evaluation.

**Comment by takuti** [comment]
The latest workflows at the moment of 50683e6968151450d3ee3b0d776b65701456c5c6 showed the following results:

|Algorithm<br />`Workflow` | Logloss | AUC |
|:---:|:---:|:---:|
|Logistic Regression<br />`predict_logress.dig`| 0.483 | 0.732 |
|Factorization Machines<br />`predict_fm.dig`| 0.687 | 0.680 |
|RandomForest<br />`predict_rf.dig`| 2.236 | 0.633 |

It is obvious that RandomForest showed surprisingly poor accuracy. Note that the number of trees is 100, and, since over-sampling rate is 1, over-sampling and calibration perform nothing. @myui Do you see any other problems in the queries except for the over-sampling and calibration stuff?

Under the very similar configuration, I built and evaluated scikit-learn's RandomForestClassifier as well, and it reasonably worked on the criteo data even with excessive feature hashing: https://github.com/takuti-sandbox/tmp/commit/bad500bdd06fc497b3b0284c3778a574e7fcd74b 

- Log Loss: 0.455317
- AUC: 0.765436

**Comment by takuti** [comment]
According to `tree_export`, many trees directly output `label=0` without no branch:

screen shot 2018-04-25 at 11 25 14

**Comment by myui** [comment]
For CTR prediction tasks, logistic regression is a great fit. It's known and reasonable for me due to the optimization criteria.

RandomForests algorithm is poor at predicting accurate probability. Logloss would always be very bad while AUC can be good. 

Factorization Machines works well where all variables are categorical. It's reasonable for me due to the optimization scheme (pair-wise interaction of quantitative variables makes thing worse).

**Comment by myui** [comment]
`According to tree_export, many trees directly output label=0 without no branch:` 

Did you use `-stratified_sampling` in RandomForests? Without it, sampled data can contain very few positive examples and then model can always predict 0 without branches.

**Comment by myui** [comment]
@takuti 

I'm not sure why Smile's RandomForests can performs worse than Scikit .. It sometimes shows better accuracy than Scikit depending on dataset. I tried to find the difference in implementations in [the past](https://github.com/haifengl/smile/issues/70) but not find exact reason.

Voting scheme be better not to use `class weight` (OOB accuracy) in certain cases.

It might be better to do [best-first split](https://github.com/apache/incubator-hivemall/blob/master/core/src/main/java/hivemall/smile/classification/DecisionTree.java#L1030) by the default. 

The splitter for sparse data might require a special treatment. Need to look into 
https://github.com/scikit-learn/scikit-learn/blob/a24c8b464d094d2c468a16ea9f8bf8d42d949f84/sklearn/tree/_splitter.pyx
https://arxiv.org/pdf/1407.7502.pdf Understanding RandomForests (thesis by Scikit-learn’s RF author)

**Comment by takuti** [comment]
Finalized. While the accuracy is not sufficient, it looks like Hivemall's issue as discussed above, and hence I hope at least workflow definition and dependent queries have no problem and be ready to merge.

File: machine-learning/ctr-prediction/docs/more.md [hunk 1]
```diff
@@ -1,8 +1,10 @@
...
+We prepared three workflows, `predict_logress.dig`, `predict_fm.dig` and `predict_rf.dif`. In order to predict CTRs from a set of features, these workflows employ machine learning algorithm **Logistic Regression**, **Factorization Machines** and **Random Forest**, respectively.
 
-We recommend you to first try `predict_logress.dig` because the number of parameters we need to configure is smaller than `predict_fm.dig`, and exporting the models to external databases such as MySQL should be easy for further integration of learnt models and production ad servers.
+We recommend you to first try `predict_logress.dig` because the number of parameters we need to configure is smaller than `predict_fm.dig`, and exporting Logistic Regression model to external databases such as MySQL is easy for further integration of learnt models and production ad servers.
+
+**Caution:** Currently, Logistic Regression is much more stable and accurate compared to Random Forest in our platform due to an implementation-related issue of underlying OSS library, [Hivemall](https://github.com/apache/incubator-hivemall).
```
Comment (takuti) [inline machine-learning/ctr-prediction/docs/more.md]: Added a note about poor accuracy of RandomForest.

Comment (myui) [reply machine-learning/ctr-prediction/docs/more.md]: Sampling scheme is different https://github.com/takuti-sandbox/tmp/commit/bad500bdd06fc497b3b0284c3778a574e7fcd74b#diff-1befd9516f7ec90288c644547380f288R32 

Also, feature pre-processing scheme is different to scikit's evaluation.
https://github.com/takuti-sandbox/tmp/commit/bad500bdd06fc497b3b0284c3778a574e7fcd74b#diff-1befd9516f7ec90288c644547380f288R29

So, we cannot conclude it's due to implementation issue. Also, we need to find/fix the implementation issue and no need to publish this article.

**Comment by myui** [comment]
@takuti used same data to your evaluation.

train
https://console.treasuredata.com/app/jobs/278595226/query (7 min, w/ stratified sampling)

predict
https://console.treasuredata.com/app/jobs/278614278/query

evaluate
https://console.treasuredata.com/app/jobs/278617694/results (wo/ model_weight)

| AUC | logloss |
|:-:|:-:|
| 0.700127115348489	| 2.1322584433581744 |

When enabling feature hashing, logloss was improved but not sure why ...

================
Uniform sampling (with feature hashing)

training 
https://console.treasuredata.com/app/jobs/275649137/query (3 hours, 34 mins!!!!!!!!! why so bad?)

prediction
https://console.treasuredata.com/app/jobs/278591587/query

evaluate
https://console.treasuredata.com/app/jobs/278592365/results

| AUC | logloss |
|:-:|:-:|
| 0.6365403661583265 | 0.5997173707991759!!!!!!!!! (why so good?) |
